test_general/api/test_app.py

```diff
@@ -549,34 +530,36 @@ def mock_response(*args, **kwargs):
     assert response.status_code == 200
 
 
-def test_partition_file_via_api_no_retryable_error_code(monkeypatch, mocker):
+def test_partition_file_via_api_not_retryable_error_code(monkeypatch, mocker):
     """
     Verify we didn't retry if the error code is not retryable
     """
     monkeypatch.setenv("UNSTRUCTURED_PARALLEL_MODE_ENABLED", "true")
     monkeypatch.setenv("UNSTRUCTURED_PARALLEL_MODE_URL", "unused")
     monkeypatch.setenv("UNSTRUCTURED_PARALLEL_MODE_THREADS", "1")
+    monkeypatch.setenv("UNSTRUCTURED_PARALLEL_MODE_RETRY_ATTEMPTS", "3")
 
-    monkeypatch.setenv("UNSTRUCTURED_PARALLEL_RETRY_ATTEMPTS", "2")
-    monkeypatch.setenv("UNSTRUCTURED_PARALLEL_RETRY_BACKOFF_TIME", "0.1")
+    remote_partition = Mock(side_effect=HTTPException(status_code=401))
 
     monkeypatch.setattr(
         requests,
         "post",
-        lambda *args, **kwargs: MockResponse(status_code=401),
+        remote_partition,
     )
-    mock_sleep = mocker.patch("time.sleep")
     client = TestClient(app)
     test_file = Path("sample-docs") / "layout-parser-paper.pdf"
 
     response = client.post(
         MAIN_API_ROUTE,
         files=[("files", (str(test_file), open(test_file, "rb"), "application/pdf"))],
-        data={"pdf_processing_mode": "parallel"},
     )
 
     assert response.status_code == 401
-    assert mock_sleep.call_count == 0
+
+    # Often page 2 will start processing before the page 1 exception is raised.
+    # So we can't assert called_once, but we can assert the count is less than it
+    # would have been if we used all retries.
+    assert remote_partition.call_count < 4
```

Comment:
why this needs retry tho? should the 401 error from page 1 be thrown immediately since its not retryable?

Reply:
Yeah, asserting on `called_once` should work like that - we call the mock, raise the error, and bail. The problem I'm having is with the ThreadpoolExecutor.map that we use. The first page raises an exception, but then before we retrieve it in the results, the second page goes through and also hits the error. When I run this locally, the call usually happens twice. Instead of accounting for timing of the threads, we can just assert that call_count is below 4 (3 retries + the original).